Feat/restaurant detail api

inhyeokjo opened the pull request:
## 🤩 개요
식당 상세 페이지 API 구현
- 사이드 메뉴와 주 메뉴를 구분해서 보여줘야 한다.
- 가격과 이름, 전체 메뉴 갯수, 주/사이드 메뉴 갯수를 보여준다.

## 🧑‍💻 작업 사항
- Controller/Service/Repositroy 구현
- Test Code 구현
- Controller Advice 구현

## 📖 참고 사항
- 기존 코드 리팩토링 필요

inhyeokjo committed:
Chore(lombok): lombok 의존성 변경

TestCode에서 lombok을 사용하지 못하는 문제 해결하기 위해서 compileOnly -> implementation 으로 변경

inhyeokjo committed:
Feat: 식당 세부 정보 조회 API 구현

inhyeokjo committed:
Test: 식당 세부 정보 조회 API 테스트

jhbaik1501 commented on build.gradle:
```diff
@@ -24,7 +24,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-validation:3.0.2'
     implementation 'org.springdoc:springdoc-openapi-starter-webmvc-ui:2.0.2'
-    compileOnly 'org.projectlombok:lombok'
+    implementation 'org.projectlombok:lombok:1.18.26'
```
버젼을 명시하신 이유가 따로 있으실까요?

inhyeokjo replied:
버전을 명시하지 않으면 의존성을 찾지 못하는 버그가 있어서 명시했습니다.
<img width="683" alt="image" src="https://user-images.githubusercontent.com/68802559/234245408-ffaed794-515d-41c6-8cc7-1a64e143eb62.png">

jhbaik1501 commented on src/main/java/com/posomo/saltit/global/aop/ControllerAdvice.java:
```diff
@@ -0,0 +1,41 @@
+	protected String noRecordExceptionHandler(NoRecordException e) {
+		log.error(e.getMessage());
+		log.error(getStackTrace(e));
+		return "해당 레코드는 존재하지 않습니다.";
```
ResponseCode와 같이 ENUM 상수로 관리하는게 좋지 않을까요?

inhyeokjo replied:
반영했습니다. ENUM을 사용하지는 않았고 그냥 상수로 빼서 관리하도록 변경했습니다. 추후에 리뷰 바랍니다!

inhyeokjo committed:
Fix(RestaurantRepository): remove EntityGraph

@Query 어노테이션을 사용해서 @EntityGraph를 사용하지 않아도 된다.

jhbaik1501 commented on src/main/java/com/posomo/saltit/controller/RestaurantController.java:
```diff
@@ -50,6 +52,12 @@ public ResponseEntity<Object> getRestaurantSummaries(@RequestBody RestaurantFilt
+
+    @Operation(summary = "식당 세부 정보 조회 api")
+    @GetMapping("/restaurant/detail/{restaurantId}")
+    public RestaurantDetailResponse getRestaurantDetail(@PathVariable("restaurantId") @NotBlank Integer restaurantId) {
```
ControllerAdvice에서 @NotBlank 오류를 잡아야할것 같네요

jhbaik1501 replied:
어차피 Integer로 받으면 @NotBlank를 붙이지 않아도 되지 않을까요?

inhyeokjo replied:
반영했습니다!

inhyeokjo committed:
Fix(Controller): @NotBlank 제거

restaurantId는 pathVariable으로 받기 떄문에 Null이나 Blank가 들어올 수 없다. 따라서 @NotBlank는 필요 없다.

jhbaik1501 commented on src/test/java/com/posomo/saltit/respository/RestaurantRepositoryTest.java:
```diff
@@ -1,4 +1,89 @@
+	@Nested
+	class findByIdWithMenus {
+		@Test
+		@DisplayName("정상흐름")
```
[레스토랑 아이디를 기준으로 메뉴를 가져오기 : 성공]
<br> 이런 식으로 쓰는게 어떨까요? 아니면 ok()라는 메소드 명을 바꾸는 것도 괜찮을 것 같아요 😊

inhyeokjo replied:
@Nested 뒤에 @DisplayName()을 추가해서 어떤 메소드를 테스트하는지 설명하겠습니다!

inhyeokjo replied:
반영했습니다!

inhyeokjo committed:
Fix(ControllerAdvice): TypeMismatchException Handle

Append Request Param Type Mismatch Exception Handler

jhbaik1501 commented on src/test/java/com/posomo/saltit/service/RestaurantServiceTest.java:
```diff
@@ -0,0 +1,108 @@
+			menus.add(new RestaurantMenu(4L, null, "sideMenu1", 2000, null, null, false));
+			menus.add(new RestaurantMenu(5L, null, "sideMenu2", 1000, null, null, false));
+			Restaurant restaurant = Restaurant.create(1L, null, "테스트 식당", null, 100, null, null, menus, null, null, null);
+			Mockito.when(restaurantRepository.findByIdWithMenus(1L)).thenReturn(Optional.of(restaurant));
```
Mockito를 import static org.mockito.Mockito.*; 빼는 건 어때보이나요?

inhyeokjo replied:
그게 맞습니다. 실수입니다. 죄송합니다.

inhyeokjo replied:
반영했습니다!

inhyeokjo committed:
Style(Mockito): static import Mockito, Assertions

inhyeokjo committed:
Refactor: @Nested Class에 @DisplayName 추가

기존 Nested Class Name이 가독성이 떨어지고, 하위 Test Method의 DisplayName이 직관적이지 않아서, @Nested가 붙은 Inner Class에 @DisplayName을 추가해서 가독성을 높힘.

inhyeokjo committed:
Feat(P6spy): add P6spy on test

테스트 환경에서 Query를 더 직관적으로 볼 수 있도록 P6spy 도입

inhyeokjo committed:
Test: 식당 상세 조회시 PathParam이 숫자가 아닌 경우 테스트

inhyeokjo committed:
Test: restaurant와 menu를 fetch join하는 쿼리 테스트

inhyeokjo committed:
Fix: 모든 연관관계를 fetch=Lazy로 설정

필요하지 않는 데이터를 모두 가져오는 것을 예방하기 위해서 LazyLoading을 Default로 둔다. 또한 restaurant와 menu를 같이 특정 테이블만 조인해서 가져오는 커스텀 JPQL 쿼리에서도 eager로 설정되어있는 다른 필드를 가져오기 위해 n+1의 문제가 여러번 발생하는 것을 확인했다.

inhyeokjo committed:
Fix: change findByIdWithMenus query

@OneToOne 이 걸려있는 필드 중에 연관관계의 주인이 아닌경우 LazyLoading이 적용되지 않아서 기존 쿼리가 생성하는 컬럼 갯수만큼 추가 쿼리를 날린다(LazyLoading이 적용되지 않는 필드 데이터에 대한 쿼리) 이를 해결하기 위해 fetch join을 추가한다.

johan1103 commented on src/main/java/com/posomo/saltit/domain/restaurant/dto/RestaurantDetailResponse.java:
```diff
@@ -0,0 +1,83 @@
+	private Classification main;
+
+	@Schema(description = "사이드 메뉴")
+	private Classification side;
```
DTO에서 사이드 메뉴와 메인 메뉴를 따로 제공하는 이유는 저희 서비스 식당 상세페이지에서 메인 메뉴량 사이드 메뉴를 다른 섹션에서 보여주기 때문일까요?

inhyeokjo replied:
넵 맞습니다

johan1103 commented on src/main/java/com/posomo/saltit/domain/restaurant/dto/RestaurantDetailResponse.java:
### 리뷰 확인 기능
<img width="161" alt="image" src="https://user-images.githubusercontent.com/71641610/234539652-a3fdb575-e410-4b1f-a6f0-96770a3b1b67.png">
요 컴포넌트를 클릭하면 웹 뷰 형식으로 다이닝코드의 리뷰 페이지 보여줘야 합니다!

다이닝 코드는 [https://m.diningcode.com/review/list.php?rid="식당 고유 rid"] 형식으로 리뷰 정보를 제공하고 있습니다.
**고래서 저희 DB에 저장되어 있는 다이닝코드의 rid값을 DTO에 담아줘야 할 것 같습니다!**

리뷰 화면 링크 예시
[https://m.diningcode.com/review/list.php?rid=LMy6BQjFbE2W](url)

 🤔  그리고 다이닝코드의 [Robots.txt](https://www.diningcode.com/robots.txt)에서 리뷰 정보는 크롤링을 권고하지 않기 때문에  따로 크롤링 하지 않았습니다!

inhyeokjo replied:
주소를 통째로 보내주는 편이 더 좋지 않을까요? 주소가 변경되었을 때 서버는 저희가 일괄 적으로 변경할 수 있지만, 안드로이드는 어플이라서 사용자가 업데이트 하지 않으면 수정되지 않으니까요!

inhyeokjo replied:
반영했습니다.

inhyeokjo committed:
refactor: 문자열을 상수로 변경

jhbaik1501 reviewed (approved):
Good
